File: tests/fixtures/projects/large_project/session_manager.py
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Any
from datetime import datetime, timedelta
import hashlib
import json
import uuid


@dataclass
class User:
    id: str
    username: str
    email: str
    created_at: datetime
    roles: List[str] = field(default_factory=list)
    preferences: Dict[str, Any] = field(default_factory=dict)


@dataclass
class Session:
    user_id: str
    token: str
    created_at: datetime
    expires_at: datetime
    last_activity: datetime


class SessionManager:
    def __init__(self, session_duration_hours: int = 24):
        self.sessions: Dict[str, Session] = {}
        self.users: Dict[str, User] = {}
        self.session_duration = timedelta(hours=session_duration_hours)

    def register_user(self, username: str, email: str, roles: Optional[List[str]] = None) -> str:
        user_id = str(uuid.uuid4())
        user = User(
            id=user_id,
            username=username,
            email=email,
            created_at=datetime.now(),
            roles=roles or ["user"],
        )
        self.users[user_id] = user
        return user_id

    def create_session(self, user_id: str) -> str:
        if user_id not in self.users:
            raise ValueError("User not found")

        token = hashlib.sha256(f"{user_id}{datetime.now().isoformat()}".encode()).hexdigest()
        now = datetime.now()
        session = Session(
            user_id=user_id,
            token=token,
            created_at=now,
            expires_at=now + self.session_duration,
            last_activity=now,
        )
        self.sessions[token] = session
        return token

    def validate_session(self, token: str) -> Optional[User]:
        if token not in self.sessions:
            return None

        session = self.sessions[token]
        if session.expires_at < datetime.now():
            del self.sessions[token]
            return None

        session.last_activity = datetime.now()
        return self.users.get(session.user_id)

    def revoke_session(self, token: str) -> bool:
        if token in self.sessions:
            del self.sessions[token]
            return True
        return False

    def cleanup_expired_sessions(self) -> int:
        now = datetime.now()
        expired = [t for t, s in self.sessions.items() if s.expires_at < now]
        for t in expired:
            del self.sessions[t]
        return len(expired)


if __name__ == "__main__":
    manager = SessionManager()

    user_id = manager.register_user(
        "alice",
        "alice@example.com",
        roles=["user", "admin"]
    )
    print(f"Registered user: {user_id}")

    token = manager.create_session(user_id)
    print(f"Session token: {token}")

    user = manager.validate_session(token)
    if user:
        print(f"Session valid for: {user.username}")

    expired_count = manager.cleanup_expired_sessions()
    print(f"Cleaned up {expired_count} expired sessions")

File: tests/fixtures/projects/large_project/session_manager.rs
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::{Duration, SystemTime};

#[derive(Debug, Clone)]
struct User {
    id: u64,
    username: String,
    email: String,
    created_at: SystemTime,
}

#[derive(Debug)]
struct Session {
    user_id: u64,
    token: String,
    expires_at: SystemTime,
}

struct UserSessionManager {
    users: Arc<Mutex<HashMap<u64, User>>>,
    sessions: Arc<Mutex<HashMap<String, Session>>>,
    session_duration: Duration,
}

impl UserSessionManager {
    fn new(session_duration_hours: u64) -> Self {
        UserSessionManager {
            users: Arc::new(Mutex::new(HashMap::new())),
            sessions: Arc::new(Mutex::new(HashMap::new())),
            session_duration: Duration::from_secs(session_duration_hours * 3600),
        }
    }

    fn register_user(&self, username: String, email: String) -> Result<u64, String> {
        let mut users = self.users.lock().map_err(|e| e.to_string())?;
        let id = (users.len() + 1) as u64;
        let user = User {
            id,
            username,
            email,
            created_at: SystemTime::now(),
        };
        users.insert(id, user);
        Ok(id)
    }

    fn create_session(&self, user_id: u64, token: String) -> Result<(), String> {
        let users = self.users.lock().map_err(|e| e.to_string())?;
        if !users.contains_key(&user_id) {
            return Err("User not found".to_string());
        }
        drop(users);

        let session = Session {
            user_id,
            token: token.clone(),
            expires_at: SystemTime::now() + self.session_duration,
        };
        let mut sessions = self.sessions.lock().map_err(|e| e.to_string())?;
        sessions.insert(token, session);
        Ok(())
    }

    fn validate_session(&self, token: &str) -> Result<u64, String> {
        let sessions = self.sessions.lock().map_err(|e| e.to_string())?;
        if let Some(session) = sessions.get(token) {
            if session.expires_at > SystemTime::now() {
                return Ok(session.user_id);
            }
        }
        Err("Invalid or expired session".to_string())
    }
}

fn main() {
    let manager = UserSessionManager::new(24);
    
    match manager.register_user("alice".to_string(), "alice@example.com".to_string()) {
        Ok(user_id) => println!("Registered user with ID: {}", user_id),
        Err(e) => println!("Failed to register user: {}", e),
    }

    match manager.create_session(1, "abc123token".to_string()) {
        Ok(_) => println!("Session created successfully"),
        Err(e) => println!("Failed to create session: {}", e),
    }

    match manager.validate_session("abc123token") {
        Ok(user_id) => println!("Session valid for user ID: {}", user_id),
        Err(e) => println!("Session validation failed: {}", e),
    }
}
